contracts/multisigs/SafeGuardFactory.sol

```diff
@@ -0,0 +1,144 @@
+error PayloadExecFailed(bytes payload);
+
+/// @dev Failed to create a Safe Guard contract for the caller.
+/// @param caller Caller address.
+error FailedCreateSafeGuard(address caller);
+
+/// @title SafeGuardFactory - Smart contract for deployment of Safe Guard contracts via the Factory functionality
+/// @author Aleksandr Kuperman - <aleksandr.kuperman@valory.xyz>
+contract SafeGuardFactory {
+    event ChangedGuard(address indexed guard);
+
+    // keccak256("guard_manager.guard.address")
+    bytes32 internal constant GUARD_STORAGE_SLOT = 0x4a204f620c8c5ccdca3fd54d003badd85ba500436a431f0cbda4f558c93c34c8;
+    // Default payload length to be processed via a delegatecall
+    // If not empty, it has to contain at least an address (20 bytes) and a call signature (4 bytes)
+    uint256 public constant DEFAULT_PAYLOAD_LENGTH = 24;
+    // The deployed factory instance address
+    address public immutable factoryInstance;
+
+    /// @dev Guard Factory constructor.
+    constructor () {
+        factoryInstance = address(this);
+    }
+
+    /// @dev Creates a safe guard that checks transactions before and after safe transaction executions.
+    /// @notice This function must be called via a delegatecall only.
+    /// @notice The function follows the standard Safe Guard creation routine where it sets the guard address
+    ///         into its dedicated guard slot followed by the event emit defined in the GuardManager contract.
+    /// @notice A multisig has to have at least three owners to set up a safe guard.
+    /// @param payload Custom payload containing at least address and a function signature for a delegatecall.
+    function createSafeGuard(bytes memory payload) external {
+        // Check for the delegatecall
+        if (address(this) == factoryInstance) {
+            revert DelegateCallOnly(factoryInstance);
+        }
+
+        // Get the number of Safe owners
+        address[] memory owners = IGnosisSafe(address(this)).getOwners();
+        // Check for the number of owners
+        uint256 numOwners = owners.length;
+        if (numOwners < 3) {
+            revert InsufficientNumberOfOwners(numOwners, 3);
+        }
+
+        // Check for the payload length
+        uint256 payloadLength = payload.length;
+        if (payloadLength > 0 && payloadLength < DEFAULT_PAYLOAD_LENGTH) {
+            revert IncorrectPayloadLength(DEFAULT_PAYLOAD_LENGTH, payloadLength);
+        }
+
+        // Check that the guard slot is empty
+        bytes32 slot = GUARD_STORAGE_SLOT;
+        address guard;
+        // solhint-disable-next-line no-inline-assembly
+        assembly {
+            guard := sload(slot)
+        }
+        if (guard != address(0)) {
+            revert SafeGuardALreadyExists(guard);
+        }
+
+        // Execute the payload, if provided
+        // The payload must be executed before the safe guard is set, as the payload could potentially break the guard
+        if (payloadLength > 0) {
```

Comment:
The custom payload is called also via a delegatecall(), as it would be called from the module manager of the GnosisSafe contract initialization.